test/RibbonTreasuryVaultWithAirSwap.ts

```diff
@@ -0,0 +1,3679 @@
+        let ownerBalanceAfter = await premiumContract.balanceOf(owner);
+
+        let performanceFeeInAsset = BigNumber.from(bid)
+          .mul(performanceFee)
+          .div(FEE_SCALING.mul(100));
+        let totalDistributed = BigNumber.from(bid).sub(performanceFeeInAsset);
+
+        assert.bnEqual(
+          userBalanceAfter.sub(userBalanceBefore),
+          totalDistributed.div(3)
+        );
+        assert.bnEqual(
+          ownerBalanceAfter.sub(ownerBalanceBefore),
+          totalDistributed.mul(2).div(3)
+        );
+
+        await expect(tx)
+          .to.emit(vault, "DistributePremium")
+          .withArgs(
+            totalDistributed,
+            [totalDistributed.div(3), totalDistributed.mul(2).div(3)],
+            [user, owner],
+            1
+          );
+
+        const userShares = await vault.shares(user);
+        await vault.connect(userSigner).initiateWithdraw(userShares);
+
+        await vault.connect(ownerSigner).commitAndClose();
+
+        await time.increaseTo((await vault.nextOptionReadyAt()).toNumber() + 1);
+
+        const secondTx = await vault.connect(keeperSigner).rollToNextOption();
+
+        await expect(secondTx)
+          .to.emit(vault, "OpenShort")
+          .withArgs(secondOptionAddress, depositAmount.mul(2), keeper);
+
+        const secondOptionContract = await getContractAt(
+          "IERC20",
+          secondOption.address
+        );
+
+        // Purchase all the oTokens
+        const secondOptionInitialOTokenBalance =
+          await secondOptionContract.balanceOf(vault.address);
+
+        // Assume premium is in stables
+        const signedOrderTwo = await signOrderForAirSwap({
+          vaultAddress: vault.address,
+          counterpartyAddress: keeperSigner.address, // We simulate keeper as the option seller
+          signerPrivateKey: keeperSignerPrivateKey,
+          sellToken: secondOption.address,
+          buyToken: premiumContract.address,
+          sellAmount: secondOptionInitialOTokenBalance.toString(),
+          buyAmount: bid,
+        });
+
+        await premiumContract
+          .connect(keeperSigner)
+          .approve(airswapContract.address, bid);
+
+        await vault.connect(keeperSigner).sellOptions(signedOrderTwo);
+
+        userBalanceBefore = await premiumContract.balanceOf(user);
+        ownerBalanceBefore = await premiumContract.balanceOf(owner);
+
+        tx = await vault.connect(keeperSigner).chargeAndDistribute();
+
+        performanceFeeInAsset = BigNumber.from(bid)
+          .mul(performanceFee)
+          .div(FEE_SCALING.mul(100));
+        totalDistributed = BigNumber.from(bid).sub(performanceFeeInAsset);
+
+        userBalanceAfter = await premiumContract.balanceOf(user);
+        ownerBalanceAfter = await premiumContract.balanceOf(owner);
+
+        assert.bnEqual(
+          userBalanceAfter.sub(userBalanceBefore),
+          BigNumber.from(0)
+        );
+        assert.bnEqual(
```

Comment:
The test for the BAL vault is failing with an off-by-one error here as shown below:
![image](https://user-images.githubusercontent.com/38809578/209654500-25dc2e0b-139b-4b7e-bcfe-7a87a89c3db8.jpeg)
I'm not too sure why this might be the case, since the BAL vault uses the same implementation as the other vaults and has the same number of decimals. I also the original tests in `RibbonTreasuryVault.ts` had some custom logic that excluded the BAL vault, but I'm not sure what was the reason for this.
![telegram-cloud-photo-size-5-6147496062563234851-y](https://user-images.githubusercontent.com/38809578/209654716-e9d98d07-0ed4-4074-82ef-f18d0d95d5ed.jpg)

Reply:
![image](https://user-images.githubusercontent.com/99089658/210249461-403f7bcb-d93c-43b8-9d85-4f802b67a0d2.png)
Part 1/3 - Added 6 console.log on the .sol file to check what was going on:

Reply:
<img width="277" alt="image" src="https://user-images.githubusercontent.com/99089658/210249670-3d37847e-cd26-4ebd-b731-2c5c1aa65845.png">
Part 2/3 - Checked the output of the console.logs - All fine for PERP, BADGER and SPELL:

`console.log("AFTER DISTRIBUTION - stableAsset.balanceOf(address(this))", stableAsset.balanceOf(address(this)));` is correctly showing 0

Reply:
<img width="439" alt="image" src="https://user-images.githubusercontent.com/99089658/210250456-48c4627e-73e7-455a-b558-242d0f6d3a7f.png">



Part 3/4 - It's not related to the token BAL itself. It could happen to any token, it just happened to be BAL by chance, due to the number rounding.

Reply:
Part 4/4 - I would suggest following the original testing script - eg:

        assert.bnLte(
          ownerBalanceAfter.sub(ownerBalanceBefore),
          totalDistributed.add(1)
        );
        assert.bnGte(
          ownerBalanceAfter.sub(ownerBalanceBefore),
          totalDistributed.sub(1)
        );

It's also ok to use a similar workaround as was previously used:
![image](https://user-images.githubusercontent.com/99089658/210251689-4044349f-75dd-40cf-b125-38bea455cd3b.png)

Reply:
Addressed in [8928237](https://github.com/ribbon-finance/ribbon-v2/pull/297/commits/892823778c487e6773542058138ce27725910b5a)